src/lang/frontend_ast.rs

```diff
@@ -0,0 +1,1262 @@
+}
+/// Individual import name (identifier or unused "_")
+#[derive(Debug, Clone, PartialEq)]
+pub enum ImportName {
+    Named(String),
+    Unused, // "_"
+}
+
+/// Batch reference (hash)
+#[derive(Debug, Clone, PartialEq)]
+pub struct BatchRef {
+    pub hash: HashHex,
+    pub span: Option<Span>,
+}
+
+/// Custom predicate definition
+#[derive(Debug, Clone, PartialEq)]
+pub struct CustomPredicateDef {
+    pub name: Identifier,
+    pub args: ArgSection,
+    pub conjunction_type: ConjunctionType,
+    pub statements: Vec<StatementTmpl>,
+    pub span: Option<Span>,
+}
+
+/// Request definition
+#[derive(Debug, Clone, PartialEq)]
+pub struct RequestDef {
+    pub statements: Vec<StatementTmpl>,
+    pub span: Option<Span>,
+}
+
+/// Argument section with public and optional private arguments
+#[derive(Debug, Clone, PartialEq)]
+pub struct ArgSection {
+    pub public_args: Vec<Identifier>,
+    pub private_args: Option<Vec<Identifier>>,
+    pub span: Option<Span>,
+}
+
+/// Conjunction type for custom predicates
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub enum ConjunctionType {
+    And,
+    Or,
+}
+
+/// Statement template: predicate call with arguments
+#[derive(Debug, Clone, PartialEq)]
+pub struct StatementTmpl {
+    pub predicate: Identifier,
+    pub args: Vec<StatementArg>,
+    pub span: Option<Span>,
+}
+
+/// Arguments that can be passed to statements
+#[derive(Debug, Clone, PartialEq)]
+pub enum StatementArg {
+    Literal(LiteralValue),
+    Identifier(Identifier),
+    AnchoredKey(AnchoredKey),
+}
+
+/// Anchored key: Var["key"] or Var.key
+#[derive(Debug, Clone, PartialEq)]
+pub struct AnchoredKey {
+    pub root: Identifier,
+    pub key: AnchoredKeyPath,
+    pub span: Option<Span>,
+}
+
+impl AnchoredKey {
+    pub fn key_str(&self) -> &str {
+        match &self.key {
+            AnchoredKeyPath::Bracket(ls) => &ls.value,
+            AnchoredKeyPath::Dot(id) => &id.name,
+        }
+    }
+}
+
+/// Key path in an anchored key
+#[derive(Debug, Clone, PartialEq)]
+pub enum AnchoredKeyPath {
+    Bracket(LiteralString), // ["key"]
+    Dot(Identifier),        // .key
+}
+
+/// Identifier (variable names, predicate names, etc.)
+#[derive(Debug, Clone, PartialEq)]
+pub struct Identifier {
+    pub name: String,
+    pub span: Option<Span>,
+}
+
+/// Hash value in hex format (0x...)
+#[derive(Debug, Clone, PartialEq)]
+pub struct HashHex {
+    pub value: String, // Includes the 0x prefix
+    pub span: Option<Span>,
+}
+
+/// All possible literal values
+#[derive(Debug, Clone, PartialEq)]
+pub enum LiteralValue {
+    Int(LiteralInt),
+    Bool(LiteralBool),
+    String(LiteralString),
+    Raw(LiteralRaw),
+    PublicKey(LiteralPublicKey),
+    SecretKey(LiteralSecretKey),
+    Array(LiteralArray),
+    Set(LiteralSet),
+    Dict(LiteralDict),
+}
+
+/// Integer literal
+#[derive(Debug, Clone, PartialEq)]
+pub struct LiteralInt {
+    pub value: i64,
+    pub raw: String, // Original string representation
```

Comment:
Why is the `raw` needed, can't we just stringify the `value` when necessary?  The same question applies to `LiteralString`